src/mixins/canvas_events.mixin.js

```diff
@@ -207,40 +212,179 @@
+    /**
+     * fire `drag` event on canvas and drag source
+     * @private
+     * @param {DragEvent} e
+     */
+    _onDragProgress: function (e) {
+      var options = {
+        e: e,
+        dragSource: this._dragSource,
+        dropTarget: this._draggedoverTarget
+      };
+      this.fire('drag', options);
+      this._dragSource && this._dragSource.fire('drag', options);
+    },
+
     /**
      * prevent default to allow drop event to be fired
+     * https://developer.mozilla.org/en-US/docs/Web/API/HTML_Drag_and_Drop_API/Drag_operations#specifying_drop_targets
      * @private
      * @param {Event} [e] Event object fired on Event.js shake
      */
-    _onDragOver: function(e) {
-      e.preventDefault();
-      var target = this._simpleEventHandler('dragover', e);
+    _onDragOver: function (e) {
+      var eventType = 'dragover',
+          target = this.findTarget(e),
+          targets = this.targets,
+          options = {
+            e: e,
+            target: target,
+            subTargets: targets,
+            dragSource: this._dragSource,
+            canDrop: false,
+            dropTarget: undefined
+          };
+      this.fire(eventType, options);
+      if (target) {
+        //  render drag selection before rendering target cursor for correct visuals
+        target.canDrop(e) && this._renderDragStartSelection();
+        target.fire(eventType, options);
+      }
+      //  make sure we fire dragenter events before dragover
       this._fireEnterLeaveEvents(target, e);
+      //  propagate the event to subtargets
+      for (var i = 0; i < targets.length; i++) {
+        target = targets[i];
+        //  accept event only if previous targets didn't
+        !e.defaultPrevented && target.canDrop(e) && this._renderDragStartSelection();
+        target.fire(eventType, options);
+      }
+      //  render drag selection in case no target accepted the event
+      !e.defaultPrevented && this._renderDragStartSelection();
+    },
+
+    /**
+     * fire `dragleave` on `dragover` targets
+     * @private
+     * @param {Event} [e] Event object fired on Event.js shake
+     */
+    _onDragEnter: function (e) {
+      var target = this.findTarget(e), targets = this.targets;
+      this.fire('dragenter', {
+        e: e,
+        target: target,
+        subTargets: targets,
+        dragSource: this._dragSource
+      });
+      //  fire dragenter on targets
+      this._fireEnterLeaveEvents(target, e);
+    },
+
+    /**
+     * fire `dragleave` on `dragover` targets
+     * @private
+     * @param {Event} [e] Event object fired on Event.js shake
+     */
+    _onDragLeave: function (e) {
+      this.fire('dragleave', {
+        e: e,
+        target: this._draggedoverTarget,
+        subTargets: this.targets,
+        dragSource: this._dragSource
+      });
+      //  fire dragleave on targets
+      this._fireEnterLeaveEvents(null, e);
+      //  clear targets
+      this.targets = [];
+      this._hoveredTargets = [];
```

Comment:
are those necessary? isn't the `this._fireEnterLeaveEvents(null, e);` fired from dragOver good enough to fire dragenter and dragleave for those cases?
When we enter a canvas with a drag operation, we also trigger dragOver.

Reply:
I added tests that show it doesn't.
There was an edge case, not sure I remember correctly, but I think it was dragging on canvas and leaving the canvas, drag would continue or something strange. Perhaps dragleave didn't fire on the drag source.

Reply:
I see there's `config.canvasEvtIn` `config.canvasEvtOut` as a param of `fireSyntheticInOutEvents` which I didn't know of but is unused in drag events.
We could change that but I'm not sure what would be the expected behavior. Should canvas fire `dragenter/leave` to comply with the spec or should it fire whenever a target changes? Current logic follows spec.

Reply:
But what specs? we don't care for real event specs. Our events are just a mean for devs to interven in things that happen. We don't need to mimic anything if not a generic flow that is expect.
If we can just pin down in a comment which is the edge case we are adding those 2 handlers for, we can solve it later.

Reply:
I mean it is expected behavior from frag/drop in HTML. That's what I was after.
And it conforms with
>  Our events are just a mean for devs to intervene in things that happen. We don't need to mimic anything if not a generic flow that is expect.

I expect canvas to fire drag enter/exit as HTML element and not for each change of target. But I don't mind supporting both. Sound reasonable.
I will try to remember what the edge case was. Why are you against solving it in the PR?

Reply:
the change of target is somehow necessary if you want to make any change when dragging over a tearget, as you probably find out with the enterEditing/exitEditing text logic.
Same is for mousedown and mouseup, you would expect them on canvas level, but you need them on object level to do anything meaningful

My point is, add only events that are strictly necessary to text dragging implementation.

Reply:
I will push a commit for canvas level event named `object:dragenter` or something like that.
I wish to leave canvas drag enter/exit, important for UI decisions.

Reply:
> are those necessary? isn't the this._fireEnterLeaveEvents(null, e); fired from dragOver good enough to fire dragenter and dragleave for those cases?
When we enter a canvas with a drag operation, we also trigger dragOver.

A better explanation:
They are necessary, it's instead of `this._simpleEventHandler('dragover', e);`